modules/sdk-core/src/account-lib/mpc/tss/ecdsa/rangeproof.ts

```diff
@@ -55,13 +55,127 @@ export async function randomCoPrimeTo(x: bigint): Promise<bigint> {
+    },
+  };
 }
 
+/**
+ * Generate iterations of Ntilde, h1, h2 discrete log proofs.
+ * @param {NTilde} ntilde Ntilde, h1, h2 to generate the proofs for.
+ * @param {bigint} x Either alpha or beta depending on whether it is a discrete log proof of
+ * h1 w.r.t h2 or h2 w.r.t h1.
+ * @param {bigint} q1 The Sophie Germain prime associated with the first safe prime p1 used to generate Ntilde.
+ * @param {bigint} q2 The Sophie Germain prime associated with the second safe prime p2 used to generate Ntilde.
+ * @param {number} iterations Number of iterations of the ZK proof
+ * (default is 128 as recommened by https://blog.verichains.io/p/vsa-2022-120-multichain-key-extraction)
+ * @returns {NTildeProof} The generated NTilde Proofs.
+ */
+export async function generateNTildeProof(
+  ntilde: NTilde,
+  x: bigint,
+  q1: bigint,
+  q2: bigint,
+  iterations = 128
+): Promise<NtildeProof> {
+  const q1MulQ2 = q1 * q2;
+  const a: bigint[] = [];
+  const alpha: bigint[] = [];
+  let msgToHash: Buffer = Buffer.concat([
+    bigIntToBufferBE(ntilde.h1),
+    bigIntToBufferBE(ntilde.h2),
+    bigIntToBufferBE(ntilde.ntilde),
+  ]);
+  for (let i = 0; i < iterations; i++) {
+    a.push(randBetween(q1MulQ2));
+    alpha.push(modPow(ntilde.h1, a[i], ntilde.ntilde));
+    msgToHash = Buffer.concat([msgToHash, bigIntToBufferBE(alpha[i], Math.ceil(bitLength(ntilde.ntilde) / 8))]);
+  }
+  const simulatedResponse = createHash('sha256').update(msgToHash).digest();
+  const t: bigint[] = [];
+  for (let i = 0; i < iterations; i++) {
+    // Get the ith bit from a buffer of bytes.
+    const ithBit = (simulatedResponse[Math.floor(i / 8)] >> (7 - (i % 8))) & 1;
+    t.push((a[i] + ((BigInt(ithBit) * x) % q1MulQ2)) % q1MulQ2);
+  }
+  return { alpha, t };
+}
+
+/**
+ * Verify discrete log proofs of h1 and h2 mod Ntilde.
+ * @param {NTilde} ntilde Ntilde, h1, h2 to generate the proofs for.
+ * @param {NtildeProof} ntildeProof Ntilde Proofs
+ * @returns {boolean} true if proof is verified, false otherwise.
+ */
+export async function verifyNtildeProof(ntilde: NTilde, ntildeProof: NtildeProof): Promise<boolean> {
+  const h1ModNtilde = ntilde.h1 % ntilde.ntilde;
+  const h2ModNtilde = ntilde.h2 % ntilde.ntilde;
+  if (h1ModNtilde === BigInt(0) || h2ModNtilde === BigInt(0)) {
+    return false;
+  }
+  if (h1ModNtilde === BigInt(1) || h2ModNtilde === BigInt(1)) {
+    return false;
+  }
+  if (h1ModNtilde === h2ModNtilde) {
+    return false;
+  }
+  let msgToHash: Buffer = Buffer.concat([
+    bigIntToBufferBE(ntilde.h1),
+    bigIntToBufferBE(ntilde.h2),
+    bigIntToBufferBE(ntilde.ntilde),
+  ]);
+  for (let i = 0; i < ntildeProof.alpha.length; i++) {
+    msgToHash = Buffer.concat([
+      msgToHash,
+      bigIntToBufferBE(ntildeProof.alpha[i], Math.ceil(bitLength(ntilde.ntilde) / 8)),
+    ]);
+  }
+  const simulatedResponse = createHash('sha256').update(msgToHash).digest();
+  for (let i = 0; i < ntildeProof.alpha.length; i++) {
```

Comment:
good point will do that.

Reply:
done

Reply:
Sorry, I updated the comment later with
"Or, may even want to stick with constant `iterations`, otherwise the verification function exposes itself to similar attack that more number of iterations want to prevent!"
please make that change.

Reply:
It's a bigger issue than array out-of-bound access.